### PR TITLE
[browser][wasm][tests] JavaScript Interop Marshal tests

### PR DESCRIPTION
**Pull request by kjpou1**

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:3]: This line is redundant. The new guidance is to use a new two-line license header.

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs:3]: This line is redundant. The new guidance is to use a new two-line license header.

File: src/mono/wasm/runtime-test.js [hunk 3]
```diff
@@ -388,4 +388,7 @@ var App = {
 			fail_exec ("Unhanded argument: " + args [0]);
 		}
 	},
-};
+	call_test_method: function (method_name, signature, args) {
```
Comment (MaximLipnin) [inline src/mono/wasm/runtime-test.js:391]: Should `signature` parameter be used anywhere?

File: src/mono/wasm/runtime-test.js [hunk 4]
```diff
@@ -388,4 +388,7 @@ var App = {
...
+		return BINDING.call_static_method("[System.Runtime.InteropServices.JavaScript.Tests]System.Runtime.InteropServices.JavaScript.Tests.HelperMarshal:"+method_name, args);
```
Comment (MaximLipnin) [inline src/mono/wasm/runtime-test.js:392]: nit: need spaces around +

File: src/mono/wasm/runtime-test.js (hunk 3 above)
Comment (kjpou1) [reply src/mono/wasm/runtime-test.js:391]: Yes, I have a change for that locally.  Thanks

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs (hunk 1 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs]: Addressed

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs (hunk 2 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs]: Addressed

File: src/mono/wasm/runtime-test.js (hunk 4 above)
Comment (kjpou1) [reply src/mono/wasm/runtime-test.js]: Addressed

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+        {
+            _f64Value = d;
+        }
+
+        internal static long _i64Value;
+        private static void InvokeLong(long l)
+        {
+            _i64Value = l;
+        }
+        internal static byte[] _byteBuffer;
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:37]: nit: need a new line before

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+        {
+            using (var bytes = new Uint8Array(buffer))
+                _byteBuffer = bytes.ToArray();
+        }
+
+        private static void MarshalByteBuffer(Uint8Array buffer)
+        {
+            _byteBuffer = buffer.ToArray();
+        }
+        internal static int[] _intBuffer;
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:48]: nit: need a new line before

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+            using (var ints = new Int32Array(buffer))
+                _intBuffer = ints.ToArray();
+        }
+
+        internal static string _stringResource;
+        private static void InvokeString(string s)
+        {
+            _stringResource = s;
+        }
+        internal static string _marshalledString;
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:60]: nit: need a new line before

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+        private static string InvokeMarshalString()
+        {
+            _marshalledString = "Hic Sunt Dracones";
+            return _marshalledString;
+        }
+        internal static object _object1;
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:66]: nit: need a new line before

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+        private static object InvokeMarshalObj()
+        {
+            _marshalledObject = new object();
+            return _marshalledObject;
+        }
+
+        internal static int _valOne, _valTwo;
+        private static void ManipulateObject(JSObject obj)
+        {
+            _valOne = (int)obj.Invoke("inc"); ;
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:90]: redundant `;`

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+        {
+            _doubleBuffer = buffer.ToArray();
+        }
+
+        private static void MarshalArrayBufferToFloat64Array(ArrayBuffer buffer)
+        {
+            using (var doubles = new Float64Array(buffer))
+                _doubleBuffer = doubles.ToArray();
+        }
+        private static void MarshalByteBufferToDoubles(ArrayBuffer buffer)
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:202]: nit: need a new line before

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+            float[] buffer = Enumerable.Repeat(3.14f, 17).ToArray();
+            obj.SetObjectProperty("typedArray", Float32Array.From(buffer));
+        }
+
+        internal static float[] _taFloat;
+        private static void GetTypedArrayFloat(JSObject obj)
+        {
+            _taFloat = ((Float32Array)obj.GetObjectProperty("typedArray")).ToArray();
+        }
+
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:292]: nit: redundant line

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs [hunk 12]
```diff
@@ -0,0 +1,571 @@
...
+            Assert.Equal(0, HelperMarshal._intBuffer[0]);
+            Assert.Equal(2, HelperMarshal._intBuffer[1]);
+            Assert.Equal(4, HelperMarshal._intBuffer[2]);
+            Assert.Equal(6, HelperMarshal._intBuffer[3]);
+        }
+
+        [Fact]
+        public static void MarshalArrayBuffer2Int2()
+        {
+
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs:64]: nit: redundant line

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs [hunk 13]
```diff
@@ -0,0 +1,571 @@
...
+            Assert.Equal("Hic Sunt Dracones", HelperMarshal._jsObjects[2]);
+            Assert.NotEqual("HIC SVNT LEONES", HelperMarshal._jsObjects[2]);
+            Assert.Equal(true, HelperMarshal._jsObjects[3]);
+        }
+
+        [Fact]
+        public static void JSObjectApply()
+        {
+            Runtime.InvokeJS(@"
+                var do_add = function(a, b) { return a + b};
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs:175]: nit: add a space before `}`

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs [hunk 14]
```diff
@@ -0,0 +1,571 @@
...
+                App.call_test_method (""UseFunction"", [ do_add ]);
+            ");
+            Assert.Equal(30, HelperMarshal._jsAddFunctionResult);
+        }
+
+        [Fact]
+        public static void JSObjectAsFunction()
+        {
+            Runtime.InvokeJS(@"
+                var do_add = function(a, b) { return a + b};
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs:185]: nit: add a space before `}`

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs [hunk 15]
```diff
@@ -0,0 +1,571 @@
...
+            Runtime.InvokeJS("App.call_test_method(\"InvokeLong\", [99])");
+            Assert.Equal(99, HelperMarshal._i64Value);
+        }
+
+        [Fact]
+        public static void MarshalArrayBuffer()
+        {
+            Runtime.InvokeJS(@"
+                var buffer = new ArrayBuffer(16);
+                App.call_test_method (""MarshalArrayBuffer"", [ buffer ]);
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs:37]: nit: in the previous test you don't have a space after a method name

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs [hunk 16]
```diff
@@ -0,0 +1,571 @@
...
+            Runtime.InvokeJS(@"
+                var obj = { };
+                App.call_test_method (""SetTypedArrayFloat"", [ obj ]);
+                App.call_test_method (""GetTypedArrayFloat"", [ obj ]);
+            ");
+            Assert.Equal(17, HelperMarshal._taFloat.Length);
+            Assert.Equal(3.14f, HelperMarshal._taFloat[0]);
+            Assert.Equal(3.14f, HelperMarshal._taFloat[HelperMarshal._taFloat.Length - 1]);
+        }
+
```
Comment (MaximLipnin) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs:534]: nit: redundant line

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs (hunk 5 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:38]: Addressed in commits

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs (hunk 6 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:50]: Addressed in commits

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs (hunk 7 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:63]: Addressed in commits

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs (hunk 8 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:70]: Addressed in commits

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs (hunk 10 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:207]: Addressed in commits

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs (hunk 11 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:297]: Addressed in commits

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs (hunk 15 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs:37]: Addressed in commits

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs (hunk 16 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs:533]: Addressed in commits

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs (hunk 9 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs]: Addressed in commits

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs (hunk 12 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs]: Addressed in commits

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs (hunk 13 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs]: Addressed in commits

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs (hunk 14 above)
Comment (kjpou1) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/MarshalTests.cs]: Addressed in commits

**Review by MaximLipnin — APPROVED** [review]

**Pull request merged**